.github/workflows/ci_pr.yml

```diff
@@ -0,0 +1,140 @@
+      run: |
+        sudo add-apt-repository ppa:deadsnakes/ppa -y
+        sudo apt-get update -y
+        sudo apt install python2.6 python2.6-dev -y
+        sudo ln -sf /usr/bin/python2.6 /usr/bin/python
+        PATH="/usr/bin/${PATH:+:${PATH}}"
+        echo $PATH
+        echo "Installed python2.6"
+        python -V
+        curl https://bootstrap.pypa.io/2.6/get-pip.py -o get-pip.py
+        sudo env "PATH=$PATH" python get-pip.py
+
+    - uses: actions/checkout@v2
+
+    - name: Install dependencies
+      run: |
+        sudo env "PATH=$PATH" pip install -r requirements.txt
+        sudo env "PATH=$PATH" pip install -r test-requirements.txt
+    
+    - name: Test with nosetests
+      run: |
+        ./ci/nosetests.sh || EXIT_CODE=$?
+        exit $EXIT_CODE
+
+  build:
+    
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+
+          - python-version: 2.7
+            PYLINTOPTS: "--rcfile=ci/2.7.pylintrc"
+            NOSEOPTS: "--verbose --with-timer"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+
+          - python-version: 3.4
+            PYLINTOPTS: "--rcfile=ci/2.7.pylintrc"
+            NOSEOPTS: "--verbose --with-timer"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+
+          - python-version: 3.6
+            PYLINTOPTS: "--rcfile=ci/3.6.pylintrc"
+            NOSEOPTS: "--verbose --with-timer"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+
+          - python-version: 3.7
+            PYLINTOPTS: "--rcfile=ci/3.6.pylintrc"
+            NOSEOPTS: "--verbose --with-timer"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+
+          - python-version: 3.8
+            PYLINTOPTS: "--rcfile=ci/3.6.pylintrc"
+            NOSEOPTS: "--verbose --with-coverage --with-xunit --xunit-file=nosetests.__uuid__.xml
+              --with-timer --cover-erase --cover-inclusive --cover-branches --cover-package=azurelinuxagent"
+            PYLINTFILES: "azurelinuxagent setup.py makepkg.py tests"
+    
+    name: "Python ${{ matrix.python-version }} Unit Tests"
+    runs-on: ubuntu-18.04
+
+    env:
+      PYLINTOPTS: ${{ matrix.PYLINTOPTS }}
+      PYLINTFILES: ${{ matrix.PYLINTFILES }}
+      NOSEOPTS: ${{ matrix.NOSEOPTS }}
+      PYTHON_VERSION: ${{ matrix.python-version }}
+
+    steps:
+    
+    - name: Checkout WALinuxAgent repo
+      uses: actions/checkout@v2
+
+    - name: Setup Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    
+    - name: Install dependencies
+      run: |
+        sudo env "PATH=$PATH" python -m pip install --upgrade pip
+        sudo env "PATH=$PATH" pip install -r requirements.txt
+        sudo env "PATH=$PATH" pip install -r test-requirements.txt
+
+    - name: Run pylint
+      run: |
+        pylint $PYLINTOPTS --jobs=0 $PYLINTFILES
+
+    - name: Test with nosetests
+      if: always()
```

Comment:
What does this check mean? If it runs always then why have this check?

Reply:
Basically even if the previous task failed, this one should always run.

Reply:
Ahh interesting. Isint that what the `fail-fast: false` is there for? Also then wouldnt we need this check for each task?

Reply:
`always()` here allows this step to run regardless of failure in prior steps. This allows our pylint to run even when the unit tests fail.

Do you think an inline comment here explaining what `always()` does would be useful? Or do you think it would be better just to link the docs to github actions at, e.g., the top of the yml file?
@pgombar, @narrieta, also interested in your viewpoints.

Reply:
nop, the fail-fast is to not fail all element in the matrix if one element fails
ex : if python 2.6 fails, and fail-fast is enabled, all other python will get automatically cancelled

But to be honest, @kevinclark19a  why do we have it here ?
If the setup part fails, the Unit-Test should not run, since the whole test will fail anyways

Usually always() is used for reporting Unit-test or code coverage failure, so even if they fail, we can still report the results

Reply:
The idea was that it is still useful to check unit test correctness, even if the linter found some errors, or vice versa. It's mainly a debuggability thing, in my opinion.

Reply:
Hmm makes sense. Then should we also add `always()` for the pylint too?

Reply:
The linter would not run in this case.
The linter have its own task, and the task is not marked as `always()`
My point here is if the setup task fails, (the one installing python / pip checking the version), we cannot know what version of python is currently running the test.

Reply:
If setup fails, both the linter and the unit tests will fail. Nosetests are after pylint, so the `always()` on nosetests is only to allow nosetests to run even if pylint has run and failed.

I think that it would be semantically better to only run nosetests if setup has succeeded but pylint has failed (or, obviously, if everything prior has succeeded), but I don't know of any simple mechanism to allow that. The `always()` on nosetests seemed like the best option, then: if setup fails, we would have pylint skipped and nosetests fail (and the failure should happen pretty quickly, as, e.g. `python` wouldn't be in the env), and that doesn't change the results of the workflow; if, however, setup succeeds but pylint fails, nosetests will still be able to run and provide info on unit test correctness. 

In that second case, even if the workflow will fail (because pylint failed), it's nice to have the info on what tests passed, if any.

Essentially, the later of the two (pylint or nosetests) needs an `always()`, but not the earlier.

Reply:
Additionally, even if the unit tests _don't_ fail in the first case, the only thing we lose is ~2 minutes of compute. The workflow will still fail because the setup already failed.

Reply:
You are correct
My bad, not sure what I was looking at, but I thought the pylint was under nosetest...
(also, maybe use `success() || failure()`, since `always()` run even when job gets  cancelled)

Reply:
>Do you think an inline comment here explaining what always() does would be useful?

Yeah for sure, more doc is good in general :)

>then: if setup fails, we would have pylint skipped and nosetests fail (and the failure should happen pretty quickly, as, e.g. python wouldn't be in the env)

Are you sure about it? Like wont the container already have some python pre-installed in it? Like my concern here is that if python-setup fails and we use `always()` then the tests might run with the pre-installed python version which is something that we dont want either.
Maybe try doing something like @Aymendje mentioned above, a combination of `success() || failure()` or something?

Reply:
> Like my concern here is that if python-setup fails and we use always() then the tests might run with the pre-installed python version which is something that we dont want either.

From my runs debugging this workflow, the failure is typically `nosetest: command not found`. But even if the wrong version of python runs the unit tests, is there any sort of impact? The workflow will still fail, because we're presupposing that the setup step failed, and we'll be able to clearly see that the right python wasn't installed. As mentioned above, my understanding is that we only lose out of ~2 min compute...

> Maybe try doing something like @Aymendje mentioned above, a combination of success() || failure() or something?

I think this would have the same issue that you're mentioning, as a failure in setup would cause `failure()` to return true, and the unit tests would run. I mean, I definitely think it's preferable so that unit tests don't get run when a job is cancelled, like @Aymendje mentioned, but I don't think it's applicable to this particular issue you're raising.

Reply:
My thoughts:

> Or do you think it would be better just to link the docs to github actions at, e.g., the top of the yml file?

I think we should have this regardless of having comments elsewhere inline in the file.

Regarding failing or continuing, I think we should treat **tests** separate from **linter**. E.g., tests can fail, but I still want the linter to run, or vice-versa.

However, if the setup step fails, then I agree it doesn't make sense to proceed with tests or linter.

Reply:
Marking this as resolved: replaced the `always()` with `steps.install-dependencies.outcome == 'success'`, which should achieve the wanted behavior of only ignoring linter failures when running the tests, and not setup failures.